include/envoy/upstream/cluster_manager.h

```diff
@@ -309,6 +372,19 @@ class ClusterManager {
   virtual const ClusterRequestResponseSizeStatNames&
   clusterRequestResponseSizeStatNames() const PURE;
   virtual const ClusterTimeoutBudgetStatNames& clusterTimeoutBudgetStatNames() const PURE;
+
+  /**
+   * Allocates an on-demand CDS API provider from configuration proto or locator.
```

Comment:
What are the semantics when we are using CDS and have both regular CDS wildcard subscription and OCDS on the same stream?

Reply:
Thanks for pointing this out. To rehash:
- Once CDS subscription enters the wildcard mode, it should ignore requests for specific cluster names.
- ODCDS currently always sends requests with specific cluster names.

So, it looks like those two currently can't work together in one stream. Now, I don't know if this ignoring of requests with specific resource names while being in wildcard mode is enforced on Envoy side or this is an expectation from the config server. I don't know what is the rationale behind this restriction, so I'd ask if lifting this restriction would be an acceptable idea.

Reply:
I think @markdroth added that restriction in https://www.envoyproxy.io/docs/envoy/latest/api-docs/xds_protocol. I also agree it would be good to be able to combine wildcard and on-demand CDS here on the same stream. Thoughts @markdroth?

Reply:
As currently defined, it's not possible to have both a wildcard subscription and subscription to specific resource names on the same stream.  Note, however, that wildcard subscriptions are intended to go away as part of the new naming scheme, so one option here would be to simply say that you can't do both unless you migrate to the new naming scheme.

If that's not feasible, we could look into making some tranport protocol change to support this.  For example, maybe the wildcard request could be represented by `resource_names: "*"` instead of leaving `resource_names` completely unset.  But we'd need to carefully look through the spec and make sure that there are no non-obvious implications of doing that.  I *think* it should be okay, because it doesn't change the two things that are really special about responses from wildcard subscriptions: the fact that the server must send all existing resources in every response and the fact that if a resource is no longer present in a response, the client can assume it's been deleted on the server.

Reply:
> As currently defined, it's not possible to have both a wildcard subscription and subscription to specific resource names on the same stream. Note, however, that wildcard subscriptions are intended to go away as part of the new naming scheme, so one option here would be to simply say that you can't do both unless you migrate to the new naming scheme.
> 

Is there a reason why both wildcard subscription and subscription to specific resource names are impossible on the same stream? I can see the conflict between SotW wildcard subscription and subscription to specific resource names (SotW update would likely delete the clusters we got on-demand). Such conflict wouldn't exist with delta wildcard subscription. I suppose I could limit the on-demand CDS to be usable only with delta gRPC.

About the naming scheme - I need to read about this, so can't say much about it. Just wanted to say that I dunno if cluster names coming from cluster_header action checking the `Host` HTTP header fit into a new naming scheme.

> If that's not feasible, we could look into making some tranport protocol change to support this. For example, maybe the wildcard request could be represented by `resource_names: "*"` instead of leaving `resource_names` completely unset. But we'd need to carefully look through the spec and make sure that there are no non-obvious implications of doing that. I _think_ it should be okay, because it doesn't change the two things that are really special about responses from wildcard subscriptions: the fact that the server must send all existing resources in every response and the fact that if a resource is no longer present in a response, the client can assume it's been deleted on the server.

Wouldn't changing the wildcard request representation to `resource_names: "*"` be a breaking change for the existing config servers?

Reply:
> Is there a reason why both wildcard subscription and subscription to specific resource names are impossible on the same stream? 

It's because currently, the way that a wildcard subscription is indicated is by leaving the `resource_names` field unset in the request.  It's not possible to have the `resource_names` field be both set (which is needed to subscribe to individual resources) and unset (needed for a wildcard subscription) in the same stream.

> I can see the conflict between SotW wildcard subscription and subscription to specific resource names (SotW update would likely delete the clusters we got on-demand). 

That's not a problem.  In SotW, the requirement for CDS and LDS is that the server must send all resources in every response.  If a single stream somehow had both a wildcard subscription and some individual resource subscriptions, then any resource matching any of those subscriptions (including the wildcard one) would need to be present in every response.  The client could tell that a resource was deleted if it received a response in which the resource was no longer present.

> About the naming scheme - I need to read about this, so can't say much about it. 

The new naming scheme completely eliminates wildcard subscriptions in favor of using a new collection resource.  So in the places where Envoy currently makes a wildcard subscription for LDS or CDS, it would instead subscribe to a specific ListenerCollection or ClusterCollection resource.

> Just wanted to say that I dunno if cluster names coming from cluster_header action checking the `Host` HTTP header fit into a new naming scheme.

From the xDS perspective, I don't see how a CDS resource subscribed to based on `cluster_header` is any different from any other on-demand CDS resource.  Envoy would probably just need some timeout such that if a cluster hasn't been used in a while, it can unsubscribe from that cluster and remove it.

I think the question of how to handle `cluster_header` is mostly orthogonal to the naming scheme discussion.

> Wouldn't changing the wildcard request representation to `resource_names: "*"` be a breaking change for the existing config servers?

Most existing servers probably don't handle non-wildcard queries for CDS at all, so they'll need changes for ODCDS anyway.  I think we could extend the protocol to say that `resource_names` unset and `resource_names: "*"` should both be considered wildcard queries, but then the actual implementation in Envoy could use the `"*"` form only when ODCDS was going to be used on the same stream.  That way, we won't break existing servers that don't need to make changes for ODCDS anyway.

Reply:
A few things:
1. If we can make this delta CDS only and solve these issues, let's do that. I rather we head towards delta xDS as the basis for the wire transport going forward and for advanced features like this, it makes sense.
2. The xRFC for the new naming scheme is up for review at https://github.com/cncf/xds/pull/6 for @krnowak FYI. We have support for glob collections and CDS in Envoy today, but this is quite experimental.

Reply:
I don't necessarily object to making ODCDS work for the incremental protocol variants only, but I don't think that actually solves the problem of having both wildcard and non-wildcard subscriptions on the same stream.  To fix that problem, we need to either make ODCDS work only with the new naming scheme, or we'd need to introduce something like the `"*"` hack.  And once we do either of those things, we can do them just as easily for SotW as we can for incremental.

Reply:
I'll have a look at the `"*"` hack.

Reply:
Could we add a `bool wildcard` field to `DeltaDiscoveryRequest`? This would have the least magic IMHO and be fully backwards compatible.

Reply:
I guess I could live with a `bool wildcard` field, but I'd prefer the `"*"` approach, for a few reasons:
- It's consistent with how glob collections work in the new naming scheme.
- It's semantically clearer, because the full set of resource locators that the client is subscribing to (one of which happens to be a wildcard) will be represented in the `resource_names` field, rather than the server having to look at two different fields to determine this.
- It doesn't add explicit support for wildcards to the transport protocol at a time when we're trying to make wildcard subscriptions go away, which is effectively just adding technical debt.

Reply:
OK, so to be clear:
* We're treating `*` as a reserved character; any existing users who are using this for a resource name are broken, but we don't expect this? We don't have this problem in `xdstp://`, since it's extremely unlikely someone has picked this schema.
* Clients are expected to only send this when ODCDS is configured. Observing a `*` indicates effectively that ODCDS is in use, distinguishing from other CDS specified discovery requests such as how gRPC works.
* If a server doesn't see `*` and does see enumerated resources in a subscription, it never sends wildcard resources.
* If a server doesn't see `*` and it sees an empty set in a subscription, it switches to wildcard mode.

Reply:
> * We're treating `*` as a reserved resource name; any existing users who are using this for a resource name are broken, but we don't expect this? We don't have this problem in `xdstp://`, since it's extremely unlikely someone has picked this schema.

My guess is that it's unlikely that anyone is actually using `*` as a resource name today, but you're certainly right that it's possible.  And I agree that if someone *is* doing that, this change would break them -- but only if they are using ODCDS.  That seems like a reasonable trade-off.

> * Clients are expected to only send this when ODCDS is configured. Observing a `*` indicates effectively that ODCDS is in use, distinguishing from other CDS specified discovery requests such as how gRPC works.

I would say that slightly differently: Observing a `*` indicates that the client is using both ODCDS *and* sending a wildcard CDS subscription.  gRPC doesn't use wildcard CDS in the first place, so it will never send a `*`.  But without the `*`, the ODCDS wire protocol is exactly the same as what gRPC does today.

To say this another way, if/when gRPC implements ODCDS, I don't think the management server will actually be able to tell the difference between that and what gRPC does today, because the wire protocol will be the same.  The only difference will be that the client will send the subscription to a particular CDS resource at a later time (i.e., it will wait until a request actually hits a route that uses a given cluster, rather than proactively doing it as soon as it receives the RouteConfiguration).

> * If a server doesn't see `*` and does see enumerated resources in a subscription, it never sends wildcard resources.

Yes.  This is exactly the same as how it works today.

> * If a server doesn't see `*` and it sees an empty set in a subscription, it switches to wildcard mode.

Almost right.  Seeing an empty set in a subscription implies wildcard mode only on the first request on a stream.  After that, an empty subscription set implies unsubscribing from the last resource of that type on the stream.  There is no way to switch into wildcard mode after the initial request for a given resource type on a stream.

Reply:
SG. Let's update https://www.envoyproxy.io/docs/envoy/latest/api-docs/xds_protocol as part of this proposal then.

Reply:
If CDS is configured and ODCDS is not configured, what initial request will contain in resource names? An empty list or `"*"`? What I'm asking here is how do we decide when to send an empty list vs. `"*"`. Is this decided statically (like for CDS we will always send `"*"`, because we will have ODCDS, but for EDS we will send an empty list, because there is no such thing like ODEDS?

I suppose that this "static" decision would be the way to go, because CDS is usually configured statically, while ODCDS might be configured much later (for example, ODCDS could be a part of some route configuration received from RDS). I'm probably wrong somewhere, so please correct me if it's the case. Thanks.

Reply:
If we require this to be explicitly configured, then we need to figure out how to handle the case where someone configures CDS to leave the field unset instead of setting it to `*` but then later configures ODCDS.  It seems better to figure this out automatically, so that it's not possible to configure this incorrectly.

Note that currently, wildcard mode is determined based on the first request for that resource type on a given stream, and there is no way to subsequently switch in or out of wildcard mode for that resource type on that stream.  So if the first CDS request on a stream has the `resource_names` field unset, then the stream is in wildcard mode for CDS resources, and the field is ignored in all subsequent CDS requests on that stream.  Conversely, if the first CDS request on a stream has the `resource_names` field set, then the stream is in non-wildcard mode for CDS resources, and the field is honored in all subsequent CDS requests on that stream; after that, if a request is sent with the field unset, that is interpretted as unsubscribing from all CDS resources instead of being interpretted as a wildcard request.  For details, see:

https://www.envoyproxy.io/docs/envoy/latest/api-docs/xds_protocol#how-the-client-specifies-what-resources-to-return
https://www.envoyproxy.io/docs/envoy/latest/api-docs/xds_protocol#unsubscribing-from-resources

I think that what we should do here is to change the spec to make it possible to switch *out* of wildcard code after the first request on a stream.  However, it will still be impossible to switch *into* wildcard mode once in non-wildcard mode, because otherwise there will be no way to unsubscribe from the last resource of a given type.  In other words:

- If the first request for a given resource type on a stream has the `resource_names` field unset, then it's in wildcard mode.
- If a subsequent request for that resource type on the stream has the `resource_names` field set, then the stream switches to non-wildcard mode.  If the client wants to retain the original wildcard subscription, then the `resource_names` field must include `*`; if it includes only other resource names but no `*`, that means to remove from the wildcard subscription and add whatever other subscriptions are specified.
- Once a stream has seen at least one non-wildcard subscription for a given resource type, a subsequent request for that same resource type with the `resource_names` field unset means to unsubscribe to everything; it does *not* indicate a wildcard subscription.  Clients wishing to unsubscribe to non-wildcard subscriptions but retain a wildcard subscription must continue to send `*` for the duration of the stream.

Note that in terms of the spec, I think we can say that servers should be prepared to accept `*` to mean wildcard even at the start of a stream, even when there are no non-wildcard subscriptions.  But we don't ever actually need to send that in Envoy, so in practice there won't be any backward-compatibility problems.

In terms of the Envoy implementation, I think we can figure out what to send automatically at the moment that we send each request.  We just need to track a bool indicating whether the stream has previously seen a non-wildcard subscription for a given resource type.  If it hasn't, then we can leave the field unset; if it has, then we send `*` instead.

FWIW, note that EDS never makes wildcard requests.  The only resource types that make wildcard requests are LDS, CDS, and (I think) SRDS.

Reply:
SG, I think your third case doesn't apply with delta xDS though.